code/modules/hydroponics/seed_extractor.dm

```diff
@@ -107,107 +109,127 @@
+			"id" = O.id,
+			"name" = O.name,
+			"variant" = O.variant,
+			"lifespan" = O.lifespan,
+			"endurance" = O.endurance,
+			"maturity" = O.maturation,
+			"production" = O.production,
+			"yield" = O.yield,
+			"potency" = O.potency,
+			"amount" = O.amount,
+		)
+		data["stored_seeds"] += list(seed_info)
+
+	data["vend_number"] = vend_amount
+	return data
+
+/obj/machinery/seed_extractor/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if(..())
+		return
 
-	var/dat = "<b>Stored seeds:</b><br>"
+	switch(action)
+		if("vend")
+			vendSeed(text2num(params["seedid"]), vend_amount)
+			add_fingerprint(usr)
+			SStgui.update_uis(src)
+		if("set_vend_num")
+			if(!length(params["vend_num"]))
+				return
+			vend_amount = clamp(text2num(params["vend_num"]), 1, MAX_DISPENSE_SEEDS)
+			add_fingerprint(usr)
 
-	if (contents.len == 0)
-		dat += "<font color='red'>No seeds</font>"
-	else
-		dat += "<table cellpadding='3' style='text-align:center;'><tr><td>Name</td><td>Variant</td><td>Lifespan</td><td>Endurance</td><td>Maturation</td><td>Production</td><td>Yield</td><td>Potency</td><td>Stock</td></tr>"
-		for (var/datum/seed_pile/O in piles)
-			dat += "<tr><td>[O.name]</td><td>[O.variant]</td><td>[O.lifespan]</td><td>[O.endurance]</td><td>[O.maturation]</td>"
-			dat += "<td>[O.production]</td><td>[O.yield]</td><td>[O.potency]</td><td>"
-			dat += "<a href='byond://?src=[src.UID()];name=[O.name];li=[O.lifespan];en=[O.endurance];ma=[O.maturation];pr=[O.production];yi=[O.yield];pot=[O.potency]'>Vend</a> ([O.amount] left)</td></tr>"
-		dat += "</table>"
-	var/datum/browser/popup = new(user, "seed_ext", name, 700, 400)
-	popup.set_content(dat)
-	popup.open()
-	return
-
-/obj/machinery/seed_extractor/Topic(href, list/href_list)
-	if(..())
-		return 1
-	usr.set_machine(src)
-
-	href_list["li"] = text2num(href_list["li"])
-	href_list["en"] = text2num(href_list["en"])
-	href_list["ma"] = text2num(href_list["ma"])
-	href_list["pr"] = text2num(href_list["pr"])
-	href_list["yi"] = text2num(href_list["yi"])
-	href_list["pot"] = text2num(href_list["pot"])
-
-	for (var/datum/seed_pile/N in piles)//Find the pile we need to reduce...
-		if (href_list["name"] == N.name && href_list["li"] == N.lifespan && href_list["en"] == N.endurance && href_list["ma"] == N.maturation && href_list["pr"] == N.production && href_list["yi"] == N.yield && href_list["pot"] == N.potency)
+/obj/machinery/seed_extractor/proc/vendSeed(seed_id, amount)
+	if(!seed_id)
+		return
+	var/datum/seed_pile/selected_pile
+	for(var/datum/seed_pile/N in piles)
+		if(N.id == seed_id)
+			N.amount -= amount
+			selected_pile = N
 			if(N.amount <= 0)
-				return
-			N.amount = max(N.amount - 1, 0)
-			if (N.amount <= 0)
 				piles -= N
-				qdel(N)
 			break
-
-	for (var/obj/T in contents)//Now we find the seed we need to vend
+	if(!selected_pile)
+		return
+	var/amount_dispensed = 0
+	for(var/obj/T in contents)
+		if(amount_dispensed >= amount)
+			break
 		var/obj/item/seeds/O = T
```

Comment:
yes, you've discovered how garbage this storage system is. I considered just QDEL'ing seeds and tracking them in piles as data rather than a stored object and making a new object when we wanted to pull them out. However this far out of scope for what I want to do with the PR as it would require an entire refactor of the seed extractor. I agree with the type check tho.